File: Locative/Lib/BackgroundBlockOperation.m
#import "BackgroundBlockOperation.h"

@interface BackgroundBlockOperation ()

@property (assign) UIBackgroundTaskIdentifier backgroundTaskIdentifier;

@end

@implementation BackgroundBlockOperation

- (id)init {
    if (self = [super init]) {
        self.automaticallyEndsBackgroundTask = YES;
    }
    return self;
}

- (void)startBackgroundTask {
    self.backgroundTaskIdentifier = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
        [self endBackgroundTask];
    }];
}

- (void)endBackgroundTask {
    if (self.backgroundTaskIdentifier != UIBackgroundTaskInvalid) {
        [[UIApplication sharedApplication] endBackgroundTask:self.backgroundTaskIdentifier];
        self.backgroundTaskIdentifier = UIBackgroundTaskInvalid;
    }
}

- (void)addExecutionBlock:(void (^)(void))block {
    __weak typeof(self) weakSelf = self;
    [super addExecutionBlock:^{
        [weakSelf startBackgroundTask];
        block();
        if (weakSelf.automaticallyEndsBackgroundTask)
            [weakSelf endBackgroundTask];
    }];
}

@end

File: Locative/Lib/BackgroundBlockOperation.swift
import Foundation

open class BackgroundBlockOperation: BlockOperation {
    var automaticallyEndsBackgroundTask = true
    var backgroundTaskIdentifier: UIBackgroundTaskIdentifier?
    
    func startBackgroundTask() {
        qualityOfService = .background
        backgroundTaskIdentifier = UIApplication.shared
            .beginBackgroundTask (expirationHandler: { [weak self] in
                self?.endBackgroundTask()
        })
    }
    
    func endBackgroundTask() {
        if let identifier = backgroundTaskIdentifier ,
            identifier != UIBackgroundTaskInvalid {
            UIApplication.shared.endBackgroundTask(identifier)
            backgroundTaskIdentifier = UIBackgroundTaskInvalid
        }
    }
    
    override open func addExecutionBlock(_ block: @escaping ()->()) {
        super.addExecutionBlock { [weak self] in
            self?.startBackgroundTask()
            block()
            if let this = self , this.automaticallyEndsBackgroundTask {
                self?.endBackgroundTask()
            }
        }
    }
}
